Replace some uses of {F,f}orall_expr by ranged-for

tautschnig opened the pull request:
In some (but not all!) cases of {F,}orall_expr we do not actually need the iterator.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue commented on src/ansi-c/expr2c.cpp:
```diff
@@ -2358,7 +2358,7 @@ std::string expr2ct::convert_side_effect_expr_function_call(
   dest+='(';
 
-  forall_expr(it, src.arguments())
+  for(auto it = src.arguments().begin(); it != src.arguments().end(); ++it)
```
Think this should probably be a separate refactor, but the only thing the iterator is being used for here is checking whether this is the first iteration, which we normally just do with a flag + for-each. Same is true for at least some of the other `for(auto it = something.begin(); it != something.end(); ++it)` loops. I was thinking it'd be nice to have something like `for(auto const &it : iterators(collection))` or `for(auto const it& : iterators(collection.begin(), collection.end())` for some cases that can't be so easily transformed too.

martin-cs commented on src/cpp/cpp_instantiate_template.cpp:
```diff
@@ -289,7 +291,9 @@ const symbolt &cpp_typecheckt::instantiate_template(
 #ifdef DEBUG
   std::cout << "A: <";
-  forall_expr(it, specialization_template_args.arguments())
+  for(auto it = specialization_template_args.arguments().begin();
```
At the risk of sounding ungrateful for the amazing clean up PRs... the ones that do still need iterators, I think they are best still with the macro.

tautschnig committed:
Replace some uses of {F,f}orall_expr by ranged-for

In some (but not all!) cases of {F,}orall_expr we do not actually need the
iterator.

tautschnig commented:
@hannes-steffenhagen-diffblue @martin-cs Thanks a lot for the feedback. I have modified to commit such as to leave in place the macro where the iterator was required. The refactoring suggested by @hannes-steffenhagen-diffblue probably should be done as well, but seems right for a separate PR.
